File: mpp-core/src/commonMain/kotlin/cc/unitmesh/devins/parser/CodeFence.kt
package cc.unitmesh.devins.parser

/**
 * 代码围栏解析器 - 核心解析逻辑
 * 从 IDEA 插件版本移植，去除 IntelliJ Platform 依赖
 */
class CodeFence(
    val languageId: String,
    val text: String,
    var isComplete: Boolean,
    val extension: String? = null
) {
    companion object {
        private var lastTxtBlock: CodeFence? = null
        val devinStartRegex = Regex("<devin>")
        val devinEndRegex = Regex("</devin>")
        val thinkingStartRegex = Regex("<thinking>")
        val thinkingEndRegex = Regex("</thinking>")
        val walkthroughStartRegex = Regex("<!--\\s*walkthrough_start\\s*-->")
        val walkthroughEndRegex = Regex("<!--\\s*walkthrough_end\\s*-->")

        fun parse(content: String): CodeFence {
            val languageRegex = Regex("\\s*```([\\w#+ ]*)")
            val lines = content.lines()

            // Check for <devin> tag first
            val devinStartMatch = devinStartRegex.find(content)
            if (devinStartMatch != null) {
                val endMatch = devinEndRegex.find(content)
                val isComplete = endMatch != null

                val devinContent = if (isComplete) {
                    content.substring(devinStartMatch.range.last + 1, endMatch!!.range.first).trim()
                } else {
                    content.substring(devinStartMatch.range.last + 1).trim()
                }

                return CodeFence("devin", devinContent, isComplete, "devin")
            }

            // Check for <thinking> tag
            val thinkingStartMatch = thinkingStartRegex.find(content)
            if (thinkingStartMatch != null) {
                val endMatch = thinkingEndRegex.find(content)
                val isComplete = endMatch != null

                val thinkingContent = if (isComplete) {
                    content.substring(thinkingStartMatch.range.last + 1, endMatch!!.range.first).trim()
                } else {
                    content.substring(thinkingStartMatch.range.last + 1).trim()
                }

                return CodeFence("thinking", thinkingContent, isComplete, "thinking")
            }

            // Check for <!-- walkthrough_start --> HTML comment
            val walkthroughStartMatch = walkthroughStartRegex.find(content)
            if (walkthroughStartMatch != null) {
                val endMatch = walkthroughEndRegex.find(content)
                val isComplete = endMatch != null

                val walkthroughContent = if (isComplete) {
                    content.substring(walkthroughStartMatch.range.last + 1, endMatch!!.range.first).trim()
                } else {
                    content.substring(walkthroughStartMatch.range.last + 1).trim()
                }

                return CodeFence("walkthrough", walkthroughContent, isComplete, "walkthrough")
            }

            var codeStarted = false
            var codeClosed = false
            var languageId: String? = null
            var codeIndentation = ""
            val codeBuilder = StringBuilder()

            for (line in lines) {
                if (!codeStarted) {
                    val trimmedLine = line.trimStart()
                    val matchResult: MatchResult? = languageRegex.find(trimmedLine)
                    if (matchResult != null) {
                        // Store the indentation to match it when looking for the closing fence
                        codeIndentation = line.substring(0, line.length - trimmedLine.length)
                        val substring = matchResult.groups[1]?.value?.trim()
                        languageId = substring
                        codeStarted = true
                    }
                } else {
                    val trimmedLine = line.trimStart()
                    if (trimmedLine == "```") {
                        codeClosed = true
                        break
                    } else {
                        codeBuilder.append(line).append("\n")
                    }
                }
            }

            val trimmedCode = codeBuilder.trim().toString()
            val extension = lookupFileExt(languageId ?: "txt")

            return if (trimmedCode.isEmpty()) {
                CodeFence(languageId ?: "", "", codeClosed, extension)
            } else {
                CodeFence(languageId ?: "", trimmedCode, codeClosed, extension)
            }
        }

        /**
         * 解析所有代码块（包括文本块和代码块）
         */
        fun parseAll(content: String): List<CodeFence> {
            val codeFences = mutableListOf<CodeFence>()
            var currentIndex = 0
            var processedContent = content
            
            if (content.contains("```devin\n")) {
                processedContent = preProcessDevinBlock(content)
            }

            // Parse <devin>, <thinking> and <!-- walkthrough_start --> tags
            val tagMatches = mutableListOf<Pair<String, MatchResult>>()
            devinStartRegex.findAll(processedContent).forEach { tagMatches.add("devin" to it) }
            thinkingStartRegex.findAll(processedContent).forEach { tagMatches.add("thinking" to it) }
            walkthroughStartRegex.findAll(processedContent).forEach { tagMatches.add("walkthrough" to it) }
            
            // Sort by position
            tagMatches.sortBy { it.second.range.first }

            for ((tagType, startMatch) in tagMatches) {
                if (startMatch.range.first >= currentIndex) {
                    if (startMatch.range.first > currentIndex) {
                        val beforeText = processedContent.substring(currentIndex, startMatch.range.first)
                        if (beforeText.trim().isNotEmpty()) {
                            parseMarkdownContent(beforeText, codeFences)
                        }
                    }

                    // Find the corresponding end tag
                    val endRegex = when (tagType) {
                        "devin" -> devinEndRegex
                        "thinking" -> thinkingEndRegex
                        "walkthrough" -> walkthroughEndRegex
                        else -> devinEndRegex
                    }
                    val endMatch = endRegex.find(processedContent, startMatch.range.last + 1)
                    val isComplete = endMatch != null

                    val tagContent = if (isComplete) {
                        processedContent.substring(startMatch.range.last + 1, endMatch!!.range.first).trim()
                    } else {
                        processedContent.substring(startMatch.range.last + 1).trim()
                    }

                    codeFences.add(CodeFence(tagType, tagContent, isComplete, tagType))
                    currentIndex = if (isComplete) {
                        endMatch!!.range.last + 1
                    } else {
                        processedContent.length
                    }
                }
            }

            if (currentIndex < processedContent.length) {
                val remainingContent = processedContent.substring(currentIndex)
                if (remainingContent.trim().isNotEmpty()) {
                    parseMarkdownContent(remainingContent, codeFences)
                }
            }

            return codeFences.filter {
                if (it.languageId == "devin" || it.languageId == "thinking" || it.languageId == "walkthrough") {
                    return@filter true
                }
                return@filter it.text.isNotEmpty()
            }
        }

        val devinRegexBlock = Regex("(?<=^|\\n)```devin\\n([\\s\\S]*?)\\n```\\n")
        val normalCodeBlock = Regex("\\s*```([\\w#+ ]*)\\n")

        fun preProcessDevinBlock(content: String): String {
            var currentContent = content

            val devinMatches = devinRegexBlock.findAll(content).toList()

            for (match in devinMatches) {
                var devinContent = match.groups[1]?.value ?: ""
                if (normalCodeBlock.find(devinContent) != null) {
                    if (!devinContent.trim().endsWith("```")) {
                        devinContent += "\n```"
                    }
                }

                val replacement = "\n<devin>\n$devinContent\n</devin>"
                currentContent = currentContent.replace(match.value, replacement)
            }

            return currentContent
        }

        private fun parseMarkdownContent(content: String, codeFences: MutableList<CodeFence>) {
            val languageRegex = Regex("\\s*```([\\w#+ ]*)")
            val lines = content.lines()

            var codeStarted = false
            var languageId: String? = null
            val codeBuilder = StringBuilder()
            val textBuilder = StringBuilder()
            var codeIndentation = ""

            for (i in lines.indices) {
                val line = lines[i]
                if (!codeStarted) {
                    // Check for code block start with any indentation
                    val trimmedLine = line.trimStart()
                    val matchResult = languageRegex.find(trimmedLine)
                    if (matchResult != null) {
                        if (textBuilder.isNotEmpty()) {
                            val textBlock = CodeFence(
                                "markdown", textBuilder.trim().toString(), true, "md"
                            )
                            lastTxtBlock = textBlock
                            codeFences.add(textBlock)
                            textBuilder.clear()
                        }

                        // Store the indentation to match it when looking for the closing fence
                        codeIndentation = line.substring(0, line.length - trimmedLine.length)
                        languageId = matchResult.groups[1]?.value?.trim()
                        codeStarted = true
                    } else {
                        textBuilder.append(line).append("\n")
                    }
                } else {
                    // Check if this line contains the closing fence
                    val trimmedLine = line.trimStart()

                    if (trimmedLine == "```") {
                        val codeContent = codeBuilder.trim().toString()
                        val codeFence = CodeFence(
                            languageId ?: "markdown",
                            codeContent,
                            true,
                            lookupFileExt(languageId ?: "md")
                        )
                        codeFences.add(codeFence)

                        codeBuilder.clear()
                        codeStarted = false
                        languageId = null
                        codeIndentation = ""
                    } else {
                        codeBuilder.append(line).append("\n")
                    }
                }
            }

            if (textBuilder.isNotEmpty()) {
                val textBlock = CodeFence("markdown", textBuilder.trim().toString(), true, "md")
                codeFences.add(textBlock)
            }

            if (codeStarted && codeBuilder.isNotEmpty()) {
                val code = codeBuilder.trim().toString()
                val codeFence = CodeFence(
                    languageId ?: "markdown", code, false, lookupFileExt(languageId ?: "md")
                )
                codeFences.add(codeFence)
            }
        }

        /**
         * 根据语言 ID 查找文件扩展名
         */
        fun lookupFileExt(languageId: String): String {
            return when (languageId.lowercase()) {
                "c#", "csharp" -> "cs"
                "c++", "cpp" -> "cpp"
                "c" -> "c"
                "java" -> "java"
                "javascript", "js" -> "js"
                "kotlin", "kt" -> "kt"
                "python", "py" -> "py"
                "ruby", "rb" -> "rb"
                "swift" -> "swift"
                "typescript", "ts" -> "ts"
                "markdown", "md" -> "md"
                "sql" -> "sql"
                "plantuml", "puml" -> "puml"
                "shell", "bash", "sh" -> "sh"
                "objective-c" -> "m"
                "objective-c++" -> "mm"
                "go" -> "go"
                "html" -> "html"
                "css" -> "css"
                "dart" -> "dart"
                "scala" -> "scala"
                "rust", "rs" -> "rs"
                "http request", "http" -> "http"
                "shell script" -> "sh"
                "devin" -> "devin"
                "json" -> "json"
                "yaml", "yml" -> "yaml"
                "xml" -> "xml"
                "toml" -> "toml"
                "dockerfile" -> "dockerfile"
                else -> languageId
            }
        }

        /**
         * 根据文件扩展名获取显示名称
         */
        fun displayNameByExt(extension: String): String {
            return when (extension.lowercase()) {
                "cs" -> "C#"
                "cpp" -> "C++"
                "c" -> "C"
                "java" -> "Java"
                "js" -> "JavaScript"
                "kt" -> "Kotlin"
                "py" -> "Python"
                "rb" -> "Ruby"
                "swift" -> "Swift"
                "ts" -> "TypeScript"
                "md" -> "Markdown"
                "sql" -> "SQL"
                "puml" -> "PlantUML"
                "sh" -> "Shell Script"
                "m" -> "Objective-C"
                "mm" -> "Objective-C++"
                "go" -> "Go"
                "html" -> "HTML"
                "css" -> "CSS"
                "dart" -> "Dart"
                "scala" -> "Scala"
                "rs" -> "Rust"
                "http" -> "HTTP Request"
                "json" -> "JSON"
                "yaml" -> "YAML"
                "xml" -> "XML"
                "toml" -> "TOML"
                else -> extension.uppercase()
            }
        }
    }
}

File: mpp-vscode/webview/src/utils/codeFence.ts
/**
 * CodeFence parser - mirrors mpp-core's CodeFence.parseAll()
 * Parses markdown content into code blocks and text blocks
 *
 * Handles:
 * - Standard markdown code fences (```)
 * - <devin> tags (converted from ```devin blocks)
 * - <thinking> tags
 * - <!-- walkthrough_start --> comments
 */

export interface CodeBlock {
  languageId: string;
  text: string;
  isComplete: boolean;
  extension?: string;
}

// Regex patterns matching mpp-core's CodeFence
const devinStartRegex = /<devin>/;
const devinEndRegex = /<\/devin>/;
const thinkingStartRegex = /<thinking>/;
const thinkingEndRegex = /<\/thinking>/;
const walkthroughStartRegex = /<!--\s*walkthrough_start\s*-->/;
const walkthroughEndRegex = /<!--\s*walkthrough_end\s*-->/;
const normalCodeBlockRegex = /\s*```([\w#+ ]*)\n/;
const languageRegex = /\s*```([\w#+ ]*)/;

/**
 * Pre-process ```devin blocks to <devin> tags
 * Matches mpp-core's preProcessDevinBlock
 */
function preProcessDevinBlock(content: string): string {
  let currentContent = content;

  // Find all ```devin blocks
  const devinMatches = [...content.matchAll(/(?:^|\n)```devin\n([\s\S]*?)\n```(?:\n|$)/g)];

  for (const match of devinMatches) {
    let devinContent = match[1] || '';

    // Check if there's an unclosed code block inside
    if (normalCodeBlockRegex.test(devinContent)) {
      if (!devinContent.trim().endsWith('```')) {
        devinContent += '\n```';
      }
    }

    const replacement = `\n<devin>\n${devinContent}\n</devin>`;
    currentContent = currentContent.replace(match[0], replacement);
  }

  return currentContent;
}

/**
 * Parse content into code blocks and text blocks
 * Handles markdown code fences (```) and special block types
 */
export function parseCodeBlocks(content: string): CodeBlock[] {
  const blocks: CodeBlock[] = [];
  let currentIndex = 0;

  // Pre-process ```devin blocks to <devin> tags
  let processedContent = content;
  if (content.includes('```devin\n')) {
    processedContent = preProcessDevinBlock(content);
  }

  // Find all special tag matches
  interface TagMatch {
    type: string;
    match: RegExpExecArray;
  }
  const tagMatches: TagMatch[] = [];

  // Find <devin> tags
  let match: RegExpExecArray | null;
  const devinRegex = new RegExp(devinStartRegex.source, 'g');
  while ((match = devinRegex.exec(processedContent)) !== null) {
    tagMatches.push({ type: 'devin', match });
  }

  // Find <thinking> tags
  const thinkingRegex = new RegExp(thinkingStartRegex.source, 'g');
  while ((match = thinkingRegex.exec(processedContent)) !== null) {
    tagMatches.push({ type: 'thinking', match });
  }

  // Find <!-- walkthrough_start --> tags
  const walkthroughRegex = new RegExp(walkthroughStartRegex.source, 'g');
  while ((match = walkthroughRegex.exec(processedContent)) !== null) {
    tagMatches.push({ type: 'walkthrough', match });
  }

  // Sort by position
  tagMatches.sort((a, b) => a.match.index - b.match.index);

  // Process each tag match
  for (const { type, match: startMatch } of tagMatches) {
    if (startMatch.index >= currentIndex) {
      // Parse content before this tag
      if (startMatch.index > currentIndex) {
        const beforeText = processedContent.substring(currentIndex, startMatch.index);
        if (beforeText.trim()) {
          parseMarkdownContent(beforeText, blocks);
        }
      }

      // Find the corresponding end tag
      const endRegex = type === 'devin' ? devinEndRegex :
                       type === 'thinking' ? thinkingEndRegex :
                       walkthroughEndRegex;

      const endMatch = endRegex.exec(processedContent.substring(startMatch.index + startMatch[0].length));
      const isComplete = endMatch !== null;

      const tagContent = isComplete
        ? processedContent.substring(
            startMatch.index + startMatch[0].length,
            startMatch.index + startMatch[0].length + endMatch!.index
          ).trim()
        : processedContent.substring(startMatch.index + startMatch[0].length).trim();

      blocks.push({
        languageId: type,
        text: tagContent,
        isComplete,
        extension: type
      });

      currentIndex = isComplete
        ? startMatch.index + startMatch[0].length + endMatch!.index + endMatch![0].length
        : processedContent.length;
    }
  }

  // Parse remaining content
  if (currentIndex < processedContent.length) {
    const remainingContent = processedContent.substring(currentIndex);
    if (remainingContent.trim()) {
      parseMarkdownContent(remainingContent, blocks);
    }
  }

  // Filter out empty blocks (except special types)
  return blocks.filter(block => {
    if (block.languageId === 'devin' || block.languageId === 'thinking' || block.languageId === 'walkthrough') {
      return true;
    }
    return block.text.trim().length > 0;
  });
}

/**
 * Parse markdown content (text and code blocks)
 */
function parseMarkdownContent(content: string, blocks: CodeBlock[]): void {
  const lines = content.split('\n');

  let codeStarted = false;
  let languageId: string | null = null;
  const codeBuilder: string[] = [];
  const textBuilder: string[] = [];

  for (const line of lines) {
    if (!codeStarted) {
      const trimmedLine = line.trimStart();
      const matchResult = languageRegex.exec(trimmedLine);

      if (matchResult) {
        // Save accumulated text
        if (textBuilder.length > 0) {
          const text = textBuilder.join('\n').trim();
          if (text) {
            blocks.push({
              languageId: 'markdown',
              text,
              isComplete: true,
              extension: 'md'
            });
          }
          textBuilder.length = 0;
        }

        languageId = matchResult[1]?.trim() || null;
        codeStarted = true;
      } else {
        textBuilder.push(line);
      }
    } else {
      const trimmedLine = line.trimStart();

      if (trimmedLine === '```') {
        // End of code block
        const codeContent = codeBuilder.join('\n').trim();
        blocks.push({
          languageId: languageId || 'markdown',
          text: codeContent,
          isComplete: true,
          extension: getExtensionForLanguage(languageId || 'md')
        });

        codeBuilder.length = 0;
        codeStarted = false;
        languageId = null;
      } else {
        codeBuilder.push(line);
      }
    }
  }

  // Add remaining text
  if (textBuilder.length > 0) {
    const text = textBuilder.join('\n').trim();
    if (text) {
      blocks.push({
        languageId: 'markdown',
        text,
        isComplete: true,
        extension: 'md'
      });
    }
  }

  // Add unclosed code block
  if (codeStarted && codeBuilder.length > 0) {
    const code = codeBuilder.join('\n').trim();
    blocks.push({
      languageId: languageId || 'markdown',
      text: code,
      isComplete: false,
      extension: getExtensionForLanguage(languageId || 'md')
    });
  }
}

/**
 * Get file extension for a language ID
 */
function getExtensionForLanguage(languageId: string): string {
  const extensionMap: Record<string, string> = {
    javascript: 'js',
    typescript: 'ts',
    python: 'py',
    java: 'java',
    kotlin: 'kt',
    rust: 'rs',
    go: 'go',
    cpp: 'cpp',
    c: 'c',
    csharp: 'cs',
    ruby: 'rb',
    php: 'php',
    swift: 'swift',
    shell: 'sh',
    bash: 'sh',
    zsh: 'sh',
    json: 'json',
    yaml: 'yaml',
    yml: 'yml',
    xml: 'xml',
    html: 'html',
    css: 'css',
    scss: 'scss',
    markdown: 'md',
    md: 'md',
    sql: 'sql',
    diff: 'diff',
    patch: 'patch'
  };
  
  return extensionMap[languageId.toLowerCase()] || languageId;
}

/**
 * Get display name for a language
 */
export function getDisplayName(languageId: string): string {
  const displayNames: Record<string, string> = {
    javascript: 'JavaScript',
    typescript: 'TypeScript',
    python: 'Python',
    java: 'Java',
    kotlin: 'Kotlin',
    rust: 'Rust',
    go: 'Go',
    cpp: 'C++',
    c: 'C',
    csharp: 'C#',
    ruby: 'Ruby',
    php: 'PHP',
    swift: 'Swift',
    shell: 'Shell',
    bash: 'Bash',
    json: 'JSON',
    yaml: 'YAML',
    xml: 'XML',
    html: 'HTML',
    css: 'CSS',
    markdown: 'Markdown',
    md: 'Markdown',
    sql: 'SQL',
    diff: 'Diff',
    patch: 'Patch',
    thinking: 'Thinking',
    walkthrough: 'Walkthrough',
    mermaid: 'Mermaid',
    devin: 'DevIn',
    plan: 'Plan'
  };
  
  return displayNames[languageId.toLowerCase()] || languageId || 'Text';
}

